### PR TITLE
Fix atlas filtering when exporting using processing algorithms

### LINKED ISSUE #44242: Export Atlas algorithms filter options are ignored when printing
# Problem

When using the export algorithm with the filter and coverage layer selected. The algorithm prints all the features exposed in the atlas not a subset of the features.

To replicate 
* Load the project in the geopackage attached
* Choose the algorithm export atlas as either image or pdf
* Set the filter in the dialog
example-settings

* Run the tool
* Notice that the output contains all the features

qgis-ver
[index.zip](https://github.com/qgis/QGIS/files/6839639/index.zip)

### PR DESCRIPTION
**Pull request by Samweli**
Fixes https://github.com/qgis/QGIS/issues/44242

Resolves issues in exporting atlas to image and PDF using processing algorithms when filter is enabled.

When exporting atlas to image or PDF with filters enabled the current QGIS master doesn't evaluate the filters causing
the exported output to contain all the coverage layers features instead of only the filtered ones.

### CODE REVIEW

File: src/core/layout/qgslayoutatlas.cpp [hunk 1]
```diff
@@ -395,6 +395,21 @@ bool QgsLayoutAtlas::endRender()
   return true;
 }
 
+void QgsLayoutAtlas::filterCoverageLayer()
+{
+  if ( mFilterFeatures && !mFilterExpression.isEmpty() )
+  {
+    mOriginalFilterExpression = mCoverageLayer->subsetString();
+    mCoverageLayer->setSubsetString( mFilterExpression );
```
Comment (nyalldawson) [inline src/core/layout/qgslayoutatlas.cpp:403]: There's a subtle difference between an expression based filter and a layer subset string. Subset strings do NOT use qgis expressions, and instead use the language of whatever filtering is available on the actual data source backend. So we can't use them interchangeably like you do here

**Review by nyalldawson — CHANGES_REQUESTED** [review]

File: src/analysis/processing/qgsalgorithmlayoutatlastoimage.cpp [hunk 2]
```diff
@@ -141,10 +141,13 @@ QVariantMap QgsLayoutAtlasToImageAlgorithm::processAlgorithm( const QVariantMap
 
     expression = parameterAsString( parameters, QStringLiteral( "FILTER_EXPRESSION" ), context );
     atlas->setFilterExpression( expression, error );
+    atlas->setFilterFeatures( !expression.isEmpty() && error.isEmpty() );
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmlayoutatlastoimage.cpp:144]: Isn't this change alone sufficient to fix the issue?

Comment (Samweli) [reply src/analysis/processing/qgsalgorithmlayoutatlastoimage.cpp:144]: This only fixes the atlas to contain a correct value of the features ids (`QgsLayoutAtlas.mFeatureIds`), when exporting the atlas, the issue https://github.com/qgis/QGIS/issues/44242 still persists.

File: src/core/layout/qgslayoutatlas.cpp (hunk 1 above)
Comment (Samweli) [reply src/core/layout/qgslayoutatlas.cpp:403]: Right.

Comment (Samweli) [reply src/core/layout/qgslayoutatlas.cpp:403]: I came up with these methods in order to have a way to filter the coverage layer before the exporting of atlas via the processing algorithms.

File: src/analysis/processing/qgsalgorithmlayoutatlastoimage.cpp (hunk 2 above)
Comment (nirvn) [reply src/analysis/processing/qgsalgorithmlayoutatlastoimage.cpp:144]: I don't think this filterCoverageLayer() / unFilterCoverageLayer() approach s right here. QgsLayoutAtlas' setFilterExpression should suffice to filter atlas features. That's the whole point of this function to begin with.

If there are other lingering issues with setFilterExpression itself, let's nail those.

also, unFilter* would have to be renamed to something like clearFilter* , but I really don't think we need this here.

**Comment by nyalldawson** [comment]
@Samweli can you confirm if the new approach works correctly?

File: src/analysis/processing/qgsalgorithmlayoutatlastoimage.cpp (hunk 2 above)
Comment (Samweli) [reply src/analysis/processing/qgsalgorithmlayoutatlastoimage.cpp:144]: @nyalldawson upon another check, I have seen that this change fixes the https://github.com/qgis/QGIS/issues/44242 issue.

**Comment by Samweli** [comment]
> @Samweli can you confirm if the new approach works correctly?

@nyalldawson yes I have confirmed the approach works correctly, I'm waiting for @NyakudyaA to also confirm.

**Comment by nyalldawson** [comment]
Great -- I'm happy with this approach! :+1:

**Review by nyalldawson — APPROVED** [review]

**Comment by nyalldawson** [comment]
(will wait for @NyakudyaA 's confirmation before we merge)

**Comment by NyakudyaA** [comment]
LGTM

**Pull request merged**